### PR TITLE
Changes:

### PR DESCRIPTION
**Pull request by dimichgh**
- cluster support
- error handling at start and during streaming
- timeout to avoid run on sockets when server doesn't respond
- retry logic
- unit tests demostrating the problems being fixed above
- added a small change to preserve npm-delegate host into headers.target attributes; this is useful when one needs to format tarball urls in couchdb show script as pointing to npm-delgate when couchdb is under firewall and only npm-delegate can access it.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,6 +10,9 @@
   "scripts": {
     "test": "mocha -R spec"
   },
+  "scripts": {
+    "test": "mocha --ui bdd --reporter spec  ./test"
```
Comment (evocateur) [inline package.json:14]: Pretty sure this is redundant, and conflicts anyway with the `scripts` block above.

File: lib/cli.js [hunk 2]
```diff
@@ -25,7 +36,24 @@ var optimist = require('optimist')
...
-(argv.secure ? https : http).createServer(
-  delegate(argv._, proxy, argv.strictssl)
-).listen(argv.port);
-console.log("npm-delegate listening on port ", argv.port);
+var server = (argv.secure ? https : http).createServer(
+  delegate(argv._, proxy, argv.strictssl, {retry: argv.retry, timeout: argv.timeout})
+);
+
+if (argv.cluster) {
+  var Cluster = require('cluster2');
```
Comment (evocateur) [inline lib/cli.js:44]: Why not use the core `cluster` module? This also doesn't really provide for management of the clustered process (stop, shutdown, restart, status, etc).

**Comment by junosuarez** [comment]
@evocateur I'm okay with including cluster functionality in the cli bootstrapper file. People who want to get more hands on can consumer `npm-delegate` programmatically using `var npmDelegate = require('npm-delegate/lib/delegate')`. `cluster2` is a good default.

@dimichgh I get an error when running as `--cluster`:

```
 jden:npm-delegate /enh1 clean $ node lib/cli -p 1337 http://localhost:5984/registry https://registry.npmjs.org --cluster
Running cluster... 
npm-delegate listening on port  1337
Running cluster... 
npm-delegate listening on port  1337
TypeError: Cannot read property 'port' of null
    at Server.<anonymous> (cluster.js:570:30)
    at Server.g (events.js:175:14)
    at Server.EventEmitter.emit (events.js:117:20)
    at net.js:1052:10
    at process._tickCallback (node.js:415:13)
Running cluster... 
npm-delegate listening on port  1337
TypeError: Cannot read property 'port' of null
    at Server.<anonymous> (cluster.js:570:30)
    at Server.g (events.js:175:14)
    at Server.EventEmitter.emit (events.js:117:20)
    at net.js:1052:10
    at process._tickCallback (node.js:415:13)
Running cluster... 
npm-delegate listening on port  1337
TypeError: Cannot read property 'port' of null
    at Server.<anonymous> (cluster.js:570:30)
    at Server.g (events.js:175:14)
    at Server.EventEmitter.emit (events.js:117:20)
    at net.js:1052:10
    at process._tickCallback (node.js:415:13)
Running cluster... 
npm-delegate listening on port  1337
TypeError: Cannot read property 'port' of null
    at Server.<anonymous> (cluster.js:570:30)
    at Server.g (events.js:175:14)
    at Server.EventEmitter.emit (events.js:117:20)
    at net.js:1052:10
    at process._tickCallback (node.js:415:13)
```

File: package.json (hunk 1 above)
Comment (dimichgh) [inline package.json:14]: yes, I agree

File: lib/cli.js (hunk 2 above)
Comment (dimichgh) [inline lib/cli.js:44]: cluster2 is an extension of cluster and addressed some of the pitfalls of cluster.

**Comment by dimichgh** [comment]
This is an error coming from core cluster and has nothing to do with cluster2.
It still works fine and one can ignore this.

**Comment by junosuarez** [comment]
I would prefer not to merge this until the error is resolved.

File: package.json (hunk 1 above)
Comment (junosuarez) [inline package.json:14]: Please update the PR without the additions in package.json:13-15

**Comment by dimichgh** [comment]
ok, I will work on fixing cluster, first.

**Comment by junosuarez** [comment]
If you like, we can split this into two PRs, so I can merge and publish the error handling, timeout, and retry features and then the cluster feature at a later date. Up to you.

**Comment by dimichgh** [comment]
Yes, let's split.
The cluster is secondary.
I will create another PR.

On Monday, October 28, 2013 5:21 PM, Jason Denizac notifications@github.com wrote:

If you like, we can split this into two PRs, so I can merge and publish the error handling, timeout, and retry features and then the cluster feature at a later date. Up to you.

**Pull request closed without merging**